### PR TITLE
feat(kilo-pass): record Google Play purchase amount, currency and tax (part 1/1)

### PR DESCRIPTION
**Pull request by iscekic**
## Changelog for users

- Google Play purchases now record the amount actually charged, the currency, and the tax withheld in `kilo_pass_store_purchases`.
- Warehouse revenue can be calculated per Google Play purchase from the recorded amount and currency.
- New purchases take the three values from the Google Play order data the platform already receives.
- A one-off run fills the three values for existing Google Play purchases from their stored order ids.
- A purchase whose Play order carries no money, or money that cannot be read, still completes and records nothing rather than a wrong amount.

## Changelog for maintainers

- `kilo_pass_store_purchases` gains nullable `amount_charged_minor_units` (integer), `currency` (text), and `tax_minor_units` (integer); migration 0250 is additive and adds check constraints for non-negative amounts, `^[A-Z]{3}$` currency, and a currency whenever an amount is present.
- The Play order request now asks for order- and line-item-level `total` and `tax`; the same call feeds verification, notifications, and the backfill.
- New `store-purchase-money.ts` converts Play `Money { currencyCode, units, nanos }` to the currency's smallest unit using its ISO exponent; it prefers the matched line item, then the only line item, then order totals, and maps anything unusable to nulls.
- The Google Play verifier attaches the money to the mapped purchase, and the single insert in `completeStoreKiloPassPurchase` persists it; App Store rows stay NULL.
- New `backfillGooglePlayPurchaseAmounts` re-reads each stored order id, updates only rows with both money columns NULL, and is idempotent and resumable via `limit`; failures carry row id, order id, and message.
- The backfill script supports `--limit`, `--dry-run`, and `--help`, and validates the Play service account before the first request so a bad credential exits non-zero instead of reporting only a failure count.
- Review first: the currency-exponent fallback for well-formed but unknown codes (treated as two decimals), and the rule that an item carrying unreadable money yields nulls instead of borrowing order totals.

## E2E proof

**[e10] ux-check: Kilo Pass purchase/checkout and account surfaces regression** — e10-surfaces.log records '/subscriptions' rendering 'Kilo Pass / KiloClaw / Coding Plans' and '/profile' rendering the Kilo Pass plans ('Starter', 'Pro', 'Expert', 'Buy now'), and e10-checkout.log records 'after buy url=https://checkout.stripe.com/c/pay/cs_test_' with 'Subscribe to Kilo Pass (Starter)' (screenshots e10-subscriptions.png, e10-profile.png, e10-checkout.png); the only error state seen, a 500 on /api/auto-routing/settings (e10-autorouting.log), is tagged [pre-existing] — no auto-routing file appears in e10-diffstat.log — and presents a 'Try again' action.

[e10] ux-check: Kilo Pass purchase/checkout and account surfaces regression — e2e-web/e10-profile.png (capture not retained)

**[e10] ux-check: Kilo Pass purchase/checkout and account surfaces regression**

[e10] ux-check: Kilo Pass purchase/checkout and account surfaces regression — e2e-web/e10-checkout.png (capture not retained)

**[e10] ux-check: Kilo Pass purchase/checkout and account surfaces regression**

![[e10] ux-check: Kilo Pass purchase/checkout and account surfaces regression — prior/e10-profile.png](https://github.com/user-attachments/assets/3719d1d1-1a52-431b-8c30-3718765788ae)

**[e10] ux-check: Kilo Pass purchase/checkout and account surfaces regression**

![[e10] ux-check: Kilo Pass purchase/checkout and account surfaces regression — prior/e10-checkout.png](https://github.com/user-attachments/assets/2661bfe5-6c5b-4dd6-868b-ad8c4b343700)

## E2E proof — log excerpts

```
[e2] ux-check: Kilo Pass purchase/checkout and account surfaces regression -> pass :: Live playwright run against the dev stack at localhost:4800 signed in as e2e-verifier@kilo.test: /subscriptions renders 'Manage your subscriptions and billing in one place.' with the Kilo Pass, KiloClaw ('New KiloClaw instances are unavailable') and Coding Plans ('MiniMax Token Plan Plus' ... 'Sold out') tabs, /profile renders 'Unlock up to 50% free credits. Checkout securely with Stripe.' and the purchase entry is present ('buynow count=3' in e2-surfaces.log), the first Buy now reached the hosted checkout shown as 'Subscribe to Kilo Pass (Starter)' (e2-surfaces.log), no visible element is new or changed because the diff touches no UI file (e2-diffstat.log), and no layout shift was measured:
```

<details><summary><code>/home/igor_kilocode_ai/.local/share/kwf/sections/req-20260917-192220-8435/e2e-web/e2-surfaces.log</code></summary>

```
Maldives
Mali
Malta
Martinique
Mauritania
Mauritius
Mayotte
Mexico
Moldova
Monaco
Mongolia
Montenegro
Montserrat
Morocco
Mozambique
Myanmar (Burma)
Namibia
Nauru
Nepal
Netherlands
New
consoleErrors: ["Failed to load resource: the server responded with a status of 401 (Unauthorized)","Failed to load resource: the server responded with a status of 404 (Not Found)","Failed to load res
5xx/4xx responses: ["401 http://localhost:4800/api/user","404 http://localhost:4800/ingest/array/fake-token/config.js","404 http://localhost:4800/ingest/array/fake-token/config?ip=0&_=1789672743291&ve
DONE
```
</details>

<details><summary><code>/home/igor_kilocode_ai/.local/share/kwf/sections/req-20260917-192220-8435/e2e-web/e2-diffstat.log</code></summary>

```
apps/web/src/lib/kilo-pass/store-subscription-completion.ts
apps/web/src/scripts/d2026-09-17_backfill-store-purchase-amounts.ts
packages/db/src/migrations/0250_kilo_pass_store_purchase_amounts.sql
packages/db/src/migrations/meta/0250_snapshot.json
packages/db/src/migrations/meta/_journal.json
packages/db/src/schema.ts
$ git diff --stat 4238fe94e..HEAD
 .../kilo-pass/google-play-notifications.test.ts    |    70 +
 apps/web/src/lib/kilo-pass/google-play-sdk.test.ts |    29 +-
 apps/web/src/lib/kilo-pass/google-play-sdk.ts      |    36 +-
 .../src/lib/kilo-pass/google-play-verifier.test.ts |    39 +
 apps/web/src/lib/kilo-pass/google-play-verifier.ts |     2 +
 .../store-purchase-money-backfill.test.ts          |   355 +
 .../lib/kilo-pass/store-purchase-money-backfill.ts |   184 +
 .../src/lib/kilo-pass/store-purchase-money.test.ts |   185 +
 apps/web/src/lib/kilo-pass/store-purchase-money.ts |   108 +
 .../store-subscription-completion.test.ts          |   138 +
 .../lib/kilo-pass/store-subscription-completion.ts |     6 +
 .../d2026-09-17_backfill-store-purchase-amounts.ts |   123 +
 .../0250_kilo_pass_store_purchase_amounts.sql      |     7 +
 packages/db/src/migrations/meta/0250_snapshot.json | 41877 +++++++++++++++++++
 packages/db/src/migrations/meta/_journal.json      |     7 +
 packages/db/src/schema.ts                          |    19 +
 16 files changed, 43180 insertions(+), 5 deletions(-)
```
</details>

<details><summary><code>/home/igor_kilocode_ai/.local/share/kwf/sections/req-20260917-192220-8435/e2e-web/e2-layoutshift.log</code></summary>

```
/subscriptions cls=0.0009420437282986111 shiftEntries=2 firstShifts=[0.0001,0.0008]
/profile cls=0.00007263183593750001 shiftEntries=1 firstShifts=[0.0001]
```
</details>

<details><summary><code>/home/igor_kilocode_ai/.local/share/kwf/sections/req-20260917-192220-8435/e2e-web/e2-autorouting.log</code></summary>

```
status=500
body(head 1200)=
```
</details>

<details><summary><code>/home/igor_kilocode_ai/.local/share/kwf/sections/req-20260917-192220-8435/e2e-web/web-setup-smoke.log</code></summary>

```
  ✘  1 [chromium] › tests/setup-smoke/profile.spec.ts:10:7 › local setup smoke › signs in with fake auth and renders the profile page (30.2s)
  1) [chromium] › tests/setup-smoke/profile.spec.ts:10:7 › local setup smoke › signs in with fake auth and renders the profile page
    Test timeout of 30000ms exceeded.
    Error: expect(locator).toBeVisible() failed
    Locator: getByRole('region', { name: 'User profile' })
    Expected: visible
    Error: element(s) not found
    Call log:
      - Expect "toBeVisible" with timeout 30000ms
      - waiting for getByRole('region', { name: 'User profile' })
      58 |
      59 |     const profileCard = page.getByRole('region', { name: 'User profile' });
    > 60 |     await expect(profileCard).toBeVisible({ timeout: 30_000 });
         |                               ^
      61 |     await expect(page.getByRole('link', { name: 'Your Profile' })).toBeVisible();
      62 |     await expect(profileCard.getByRole('button', { name: 'Edit profile' })).toBeVisible();
      63 |     await expect(profileCard.getByText(testEmail, { exact: true })).toBeVisible();
        at /home/igor_kilocode_ai/.local/share/kwf/wt/req-20260917-192220-8435/apps/web/tests/setup-smoke/profile.spec.ts:60:31
    attachment #1: screenshot (image/png) ──────────────────────────────────────────────────────────
    test-results/setup-smoke/profile-local-setup-smoke--f6e7e-nd-renders-the-profile-page-chromium/test-failed-1.png
    ────────────────────────────────────────────────────────────────────────────────────────────────
    Error Context: test-results/setup-smoke/profile-local-setup-smoke--f6e7e-nd-renders-the-profile-page-chromium/error-context.md
  1 failed
```
</details>

<details>
<summary>Owner request for the stack</summary>

> Add the actual amount charged, the currency, and the tax to our Google Play purchase records in kilo_pass_store_purchases. Today those records cannot tell the warehouse what a purchase really earned: we need the real amount charged, the currency it was charged in, and the tax withheld, recorded per Google Play purchase alongside what is already stored, so revenue can be calculated in the warehouse. Find the existing table and the Google Play purchase write path, extend the record with the three values, populate them for new purchases from the Google Play purchase data we already receive, and backfill what can be backfilled from the stored receipts. Prove it end to end with a real Google Play purchase record written and read back showing the amount, the currency and the tax.

</details>

## PR stack (merge bottom to top)
- part 1/1: #6282 (`kwf/req-20260917-192220-8435-l1`) ← this PR

### CODE REVIEW

File: apps/web/src/lib/kilo-pass/store-purchase-money.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  if (!currencyCode) return null;
+  const exponent = currencyExponent(currencyCode);
+  if (exponent === null) return null;
+
+  const units = money.units == null ? 0 : Number(money.units);
+  const nanos = money.nanos == null ? 0 : Number(money.nanos);
+  if (!Number.isFinite(units) || !Number.isFinite(nanos)) return null;
+
+  const minorUnits = Math.round(units * 10 ** exponent + nanos / 10 ** (9 - exponent));
+  return Number.isFinite(minorUnits) ? minorUnits : null;
```
[the remark this answers is not in the document]
Comment (iscekic) [reply apps/web/src/lib/kilo-pass/store-purchase-money.ts]: bot: Accepted. Pushed 1173234 (kwf `kwf-fix-review-7dcb`).

- `apps/web/src/lib/kilo-pass/store-purchase-money.ts:52` — accepted: a negative converted amount now maps to null, and a negative tax alone leaves the amount in place, so the non-negative check constraint can no longer reject the insert and fail the whole purchase completion; unit and notification tests cover both.

File: apps/web/src/lib/kilo-pass/store-purchase-money-backfill.ts [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+      id: kilo_pass_store_purchases.id,
+      productId: kilo_pass_store_purchases.product_id,
+      providerTransactionId: kilo_pass_store_purchases.provider_transaction_id,
+    })
+    .from(kilo_pass_store_purchases)
+    .where(
+      and(
+        eq(kilo_pass_store_purchases.payment_provider, KiloPassPaymentProvider.GooglePlay),
+        isNotNull(kilo_pass_store_purchases.provider_transaction_id),
+        isNull(kilo_pass_store_purchases.amount_charged_minor_units),
```
[the remark this answers is not in the document]
Comment (iscekic) [reply apps/web/src/lib/kilo-pass/store-purchase-money-backfill.ts:149]: bot: Accepted. Pushed 1173234 (kwf `kwf-fix-review-7dcb`).

- `apps/web/src/lib/kilo-pass/store-purchase-money-backfill.ts:137` — accepted: the batch excludes rows already marked attempted, and a no-money order marks its row attempted, so `--limit` runs converge instead of re-selecting it forever; a failed lookup stays unmarked and is retried on the next run.

File: apps/web/src/lib/kilo-pass/store-purchase-money.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  if (itemMoney.amountChargedMinorUnits !== null || itemMoney.taxMinorUnits !== null) {
+    return itemMoney;
+  }
+  if (item && (item.total != null || item.tax != null)) {
+    // The item carried money we could not interpret; the order totals would
+    // attribute other items' money to this product, so stay empty instead.
+    return noMoney();
+  }
+  if (!item) return noMoney();
+  return moneyPair(order.total, order.tax);
```
[the remark this answers is not in the document]
Comment (iscekic) [reply apps/web/src/lib/kilo-pass/store-purchase-money.ts:162]: bot: Accepted. Pushed b34ca2a (kwf `kwf-fix-review-91a7`).

- `apps/web/src/lib/kilo-pass/store-purchase-money.ts:111` — accepted: the order-level fallback ran for multi-line orders too, giving one product the whole order's money; the fallback now applies only when the order has a single line item and records nulls otherwise.

File: apps/web/src/lib/kilo-pass/store-purchase-money.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+ * currency's exponent (2 for USD, 0 for JPY, 3 for KWD, ...).
+ */
+function currencyExponent(currencyCode: string): number | null {
+  if (!ISO_4217_CURRENCY_CODE.test(currencyCode)) return null;
+  try {
+    return (
+      new Intl.NumberFormat('en-US', {
+        style: 'currency',
+        currency: currencyCode,
+      }).resolvedOptions().maximumFractionDigits ?? DEFAULT_CURRENCY_EXPONENT
```
[the remark this answers is not in the document]
Comment (iscekic) [reply apps/web/src/lib/kilo-pass/store-purchase-money.ts]: bot: Accepted. Pushed b34ca2a (kwf `kwf-fix-review-91a7`).

- `apps/web/src/lib/kilo-pass/store-purchase-money.ts:25` — accepted: the exponent came from `Intl.NumberFormat` ICU/CLDR display digits, which report no decimals for HUF and IDR though ISO 4217 gives them two; it now comes from an explicit table of the ISO 4217 non-two exponents, with a two-decimal fallback for a well-formed unknown code and null for a malformed one.

File: apps/web/src/lib/kilo-pass/store-purchase-money.ts [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+  const units = money.units == null ? 0 : Number(money.units);
+  const nanos = money.nanos == null ? 0 : Number(money.nanos);
+  if (!Number.isFinite(units) || !Number.isFinite(nanos)) return null;
+
+  const minorUnits = Math.round(units * 10 ** exponent + nanos / 10 ** (9 - exponent));
+  // A negative amount cannot be stored — the column's check constraint rejects
+  // it and the whole insert would fail — so treat it as unusable, like any other
+  // money we cannot record.
+  if (!Number.isFinite(minorUnits) || minorUnits < 0) return null;
```
Comment (pandemicsyn) [inline apps/web/src/lib/kilo-pass/store-purchase-money.ts]: bot: A finite converted amount can still overflow the PostgreSQL `integer` columns and fail purchase completion.

Suggested fix: Validate the final converted value before returning it: require `Number.isSafeInteger(minorUnits)`, `minorUnits >= 0`, and `minorUnits <= 2_147_483_647` (or change the schema deliberately to `bigint` and use a precision-safe representation). Apply the same upper-bound/safe-integer rule in `nonNegativeIntegerOrNull` for defense in depth. For example, `{ currencyCode: 'USD', units: '30000000', nanos: 0 }` currently converts to `3_000_000_000`; it passes this function and the backfill sanitizer, but cannot fit the new PostgreSQL `integer` fields. On the live path it reaches the transaction insert and prevents the purchase from completing; in the backfill it makes the update throw and aborts the remaining batch. Add regression tests for both paths that assert such values become NULL/retired rather than failing the purchase or batch.

**Review by pandemicsyn — APPROVED** [review]

File: apps/web/src/lib/kilo-pass/store-purchase-money.ts (hunk 5 above)
Comment (iscekic) [reply apps/web/src/lib/kilo-pass/store-purchase-money.ts]: bot: Accepted. Pushed 4cb7e99 (kwf `kwf-fix-review-95d5`).

- `apps/web/src/lib/kilo-pass/store-purchase-money.ts:86` — accepted: a finite converted amount could exceed the PostgreSQL `integer` maximum (USD 30,000,000 converts to `3_000_000_000`) and fail the insert. The converter now requires a non-negative safe integer no greater than `MAX_STORABLE_MINOR_UNITS` (2,147,483,647); the backfill's `nonNegativeIntegerOrNull` applies the same bound, and out-of-range money records nothing like other unusable money.
